File: deepeval/metrics/misuse/misuse.py
from typing import List, Optional, Type, Union

from deepeval.metrics import BaseMetric
from deepeval.test_case import (
    LLMTestCase,
    LLMTestCaseParams,
)
from deepeval.metrics.indicator import metric_progress_indicator
from deepeval.models import DeepEvalBaseLLM
from deepeval.utils import get_or_create_event_loop, prettify_list
from deepeval.metrics.utils import (
    construct_verbose_logs,
    trimAndLoadJson,
    check_llm_test_case_params,
    initialize_model,
)
from deepeval.metrics.misuse.template import MisuseTemplate
from deepeval.metrics.misuse.schema import *
from deepeval.metrics.api import metric_data_manager


class MisuseMetric(BaseMetric):
    _required_params: List[LLMTestCaseParams] = [
        LLMTestCaseParams.INPUT,
        LLMTestCaseParams.ACTUAL_OUTPUT,
    ]

    def __init__(
        self,
        domain: str,  # Required parameter - no defaults
        threshold: float = 0.5,
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        include_reason: bool = True,
        async_mode: bool = True,
        strict_mode: bool = False,
        verbose_mode: bool = False,
        evaluation_template: Type[MisuseTemplate] = MisuseTemplate,
    ):
        if not domain or len(domain.strip()) == 0:
            raise ValueError("domain must be specified and non-empty")

        self.domain = domain.strip().lower()
        self.threshold = 0 if strict_mode else threshold
        self.model, self.using_native_model = initialize_model(model)
        self.evaluation_model = self.model.get_model_name()
        self.include_reason = include_reason
        self.async_mode = async_mode
        self.strict_mode = strict_mode
        self.verbose_mode = verbose_mode
        self.evaluation_template = evaluation_template

    def measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
        _log_metric_to_confident: bool = True,
    ) -> float:

        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(
                        test_case,
                        _show_indicator=False,
                        _in_component=_in_component,
                        _log_metric_to_confident=_log_metric_to_confident,
                    )
                )
            else:
                self.misuses: List[str] = self._generate_misuses(
                    test_case.actual_output
                )
                self.verdicts: List[MisuseVerdict] = self._generate_verdicts()
                self.score = self._calculate_score()
                self.reason = self._generate_reason()
                self.success = self.score <= self.threshold
                self.verbose_logs = construct_verbose_logs(
                    self,
                    steps=[
                        f"Misuses:\n{prettify_list(self.misuses)}",
                        f"Verdicts:\n{prettify_list(self.verdicts)}",
                        f"Score: {self.score}\nReason: {self.reason}",
                    ],
                )
                if _log_metric_to_confident:
                    metric_data_manager.post_metric_if_enabled(
                        self, test_case=test_case
                    )

            return self.score

    async def a_measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
        _log_metric_to_confident: bool = True,
    ) -> float:

        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self,
            async_mode=True,
            _show_indicator=_show_indicator,
            _in_component=_in_component,
        ):
            self.misuses: List[str] = await self._a_generate_misuses(
                test_case.actual_output
            )
            self.verdicts: List[MisuseVerdict] = (
                await self._a_generate_verdicts()
            )
            self.score = self._calculate_score()
            self.reason = await self._a_generate_reason()
            self.success = self.score <= self.threshold
            self.verbose_logs = construct_verbose_logs(
                self,
                steps=[
                    f"Misuses:\n{prettify_list(self.misuses)}",
                    f"Verdicts:\n{prettify_list(self.verdicts)}",
                    f"Score: {self.score}\nReason: {self.reason}",
                ],
            )
            if _log_metric_to_confident:
                metric_data_manager.post_metric_if_enabled(
                    self, test_case=test_case
                )
            return self.score

    async def _a_generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        misuses = []
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                misuses.append(verdict.reason)

        prompt: dict = self.evaluation_template.generate_reason(
            misuse_violations=misuses,
            score=format(self.score, ".2f"),
        )

        if self.using_native_model:
            res, cost = await self.model.a_generate(
                prompt, schema=MisuseScoreReason
            )
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: MisuseScoreReason = await self.model.a_generate(
                    prompt, schema=MisuseScoreReason
                )
                return res.reason
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def _generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        misuses = []
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                misuses.append(verdict.reason)

        prompt: dict = self.evaluation_template.generate_reason(
            misuse_violations=misuses,
            score=format(self.score, ".2f"),
        )

        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=MisuseScoreReason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: MisuseScoreReason = self.model.generate(
                    prompt, schema=MisuseScoreReason
                )
                return res.reason
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    async def _a_generate_verdicts(self) -> List[MisuseVerdict]:
        if len(self.misuses) == 0:
            return []

        verdicts: List[MisuseVerdict] = []
        prompt = self.evaluation_template.generate_verdicts(
            misuses=self.misuses, domain=self.domain
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Verdicts)
            self.evaluation_cost += cost
            verdicts = [item for item in res.verdicts]
            return verdicts
        else:
            try:
                res: Verdicts = await self.model.a_generate(
                    prompt, schema=Verdicts
                )
                verdicts = [item for item in res.verdicts]
                return verdicts
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                verdicts = [MisuseVerdict(**item) for item in data["verdicts"]]
                return verdicts

    def _generate_verdicts(self) -> List[MisuseVerdict]:
        if len(self.misuses) == 0:
            return []

        verdicts: List[MisuseVerdict] = []
        prompt = self.evaluation_template.generate_verdicts(
            misuses=self.misuses, domain=self.domain
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Verdicts)
            self.evaluation_cost += cost
            verdicts = [item for item in res.verdicts]
            return verdicts
        else:
            try:
                res: Verdicts = self.model.generate(prompt, schema=Verdicts)
                verdicts = [item for item in res.verdicts]
                return verdicts
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                verdicts = [MisuseVerdict(**item) for item in data["verdicts"]]
                return verdicts

    async def _a_generate_misuses(self, actual_output: str) -> List[str]:
        prompt = self.evaluation_template.generate_misuses(
            actual_output=actual_output, domain=self.domain
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Misuses)
            self.evaluation_cost += cost
            return res.misuses
        else:
            try:
                res: Misuses = await self.model.a_generate(
                    prompt, schema=Misuses
                )
                return res.misuses
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["misuses"]

    def _generate_misuses(self, actual_output: str) -> List[str]:
        prompt = self.evaluation_template.generate_misuses(
            actual_output=actual_output, domain=self.domain
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Misuses)
            self.evaluation_cost += cost
            return res.misuses
        else:
            try:
                res: Misuses = self.model.generate(prompt, schema=Misuses)
                return res.misuses
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["misuses"]

    def _calculate_score(self) -> float:
        number_of_verdicts = len(self.verdicts)
        if number_of_verdicts == 0:
            return 0

        misuse_count = 0
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                misuse_count += 1

        score = misuse_count / number_of_verdicts
        return 1 if self.strict_mode and score > self.threshold else score

    def is_successful(self) -> bool:
        if self.error is not None:
            self.success = False
        else:
            try:
                self.success = self.score <= self.threshold
            except:
                self.success = False
        return self.success

    @property
    def __name__(self):
        return "Misuse"

File: typescript/src/metrics/misuse/misuse.ts
import { BaseMetric } from "../base-metrics";
import { LLMTestCase, SingleTurnParams } from "../../test-case";
import { DeepEvalBaseLLM } from "../../models";
import { resolveTemplate } from "../../templates";
import {
  initializeModel,
  generateWithSchema,
  checkSingleTurnParams,
  constructVerboseLogs,
  prettifyList,
} from "../utils";
import {
  MisusesSchema,
  VerdictsSchema,
  MisuseScoreReasonSchema,
  type MisuseVerdict,
} from "./schema";

const TEMPLATE_CLASS = "MisuseMetric";

export interface MisuseMetricOptions {
  /** The chatbot's domain (e.g. "financial advisor"). Required. */
  domain: string;
  threshold?: number;
  model?: DeepEvalBaseLLM | string;
  includeReason?: boolean;
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
}

/**
 * Misuse — is the `actualOutput` used outside its intended `domain`? Extract
 * misuse instances, judge each, then score = misuse / total. **Lower is better**
 * (`success = score <= threshold`).
 */
export class MisuseMetric extends BaseMetric {
  misuses: string[] = [];
  verdicts: MisuseVerdict[] = [];
  private readonly domain: string;

  constructor(options: MisuseMetricOptions) {
    const strictMode = options.strictMode ?? false;
    super(strictMode ? 0 : (options.threshold ?? 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      includeReason: options.includeReason ?? true,
      showIndicator: options.showIndicator,
    });
    this.requiredParams = [
      SingleTurnParams.INPUT,
      SingleTurnParams.ACTUAL_OUTPUT,
    ];
    this.domain = options.domain.trim().toLowerCase();
    const { model, usingNativeModel } = initializeModel(options.model);
    this.model = model;
    this.usingNativeModel = usingNativeModel;
    this.evaluationModel = this.model.getModelName();
  }

  async measure(testCase: LLMTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      checkSingleTurnParams(testCase, this.requiredParams, this);
      this.evaluationCost = this.usingNativeModel ? 0 : undefined;

      this.misuses = await this.generateMisuses(testCase.actualOutput);
      this.verdicts = await this.generateVerdicts();
      this.score = this.calculateScore();
      this.reason = await this.generateReason();
      this.success = this.score <= this.threshold;

      this.verboseLogs = constructVerboseLogs(this, [
        `Misuses:\n${prettifyList(this.misuses)}`,
        `Verdicts:\n${prettifyList(this.verdicts)}`,
        `Score: ${this.score}\nReason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  private async generateMisuses(actualOutput: string): Promise<string[]> {
    const prompt = resolveTemplate("metrics", TEMPLATE_CLASS, "generate_misuses", {
      actual_output: actualOutput,
      domain: this.domain,
    });
    const { misuses } = await generateWithSchema(this, prompt, MisusesSchema);
    return misuses;
  }

  private async generateVerdicts(): Promise<MisuseVerdict[]> {
    if (this.misuses.length === 0) return [];
    const prompt = resolveTemplate("metrics", TEMPLATE_CLASS, "generate_verdicts", {
      misuses: this.misuses,
      domain: this.domain,
    });
    const { verdicts } = await generateWithSchema(this, prompt, VerdictsSchema);
    return verdicts;
  }

  private async generateReason(): Promise<string | undefined> {
    if (!this.includeReason) return undefined;
    const misuseViolations = this.verdicts
      .filter((v) => v.verdict.trim().toLowerCase() === "yes")
      .map((v) => v.reason);
    const prompt = resolveTemplate("metrics", TEMPLATE_CLASS, "generate_reason", {
      misuse_violations: misuseViolations,
      score: (this.score ?? 0).toFixed(2),
    });
    const { reason } = await generateWithSchema(
      this,
      prompt,
      MisuseScoreReasonSchema,
    );
    return reason;
  }

  private calculateScore(): number {
    const total = this.verdicts.length;
    if (total === 0) return 0;
    const misuseCount = this.verdicts.filter(
      (v) => v.verdict.trim().toLowerCase() === "yes",
    ).length;
    const score = misuseCount / total;
    return this.strictMode && score > this.threshold ? 1 : score;
  }

  isSuccessful(): boolean {
    const ok = this.error == null && (this.score ?? 1) <= this.threshold;
    this.success = ok;
    return ok;
  }

  get name(): string {
    return "Misuse";
  }
}
